Put implementation for async cache that uses callback method

sphill99 opened the pull request:
This is the put methods implementation for the new async cache. Instead of using a basic output stream, we add the data to a bytebuffer, then use an asynchronous write to the specified file. The write header method works differently from before, as it just puts the information into the buffer, and then writes everything to the file all at once.

There were some non-static methods in DiskBasedCache that I made into static methods and moved them to the utility class. 

In the old read/write methods, I got rid of warnings where we were shifting a value by 0 bits. Can change back if that's preferred.

sophiahaim commented on src/main/java/com/android/volley/toolbox/DiskBasedCacheUtility.java:
```diff
@@ -37,6 +42,85 @@ static String getFilenameForKey(String key) {
+    }
+
+    /**
+     * Prunes the .
```
I think you missed part of the comment here

sophiahaim commented on src/main/java/com/android/volley/toolbox/DiskBasedCacheUtility.java:
```diff
@@ -37,6 +42,85 @@ static String getFilenameForKey(String key) {
+     * @param maxCacheSizeInBytes Maximum size of the cache.
+     * @param entries Map of the entries in the cache.
+     * @param rootDirectorySupplier The supplier for the root directory to use for the cache.
+     * @return A long to update the totalSize.
```
nit: This reads a bit strange. Maybe:
@return The updated totalSize

Same for #putEntry below

sophiahaim commented on src/main/java/com/android/volley/toolbox/DiskBasedCacheUtility.java:
```diff
@@ -37,6 +42,85 @@ static String getFilenameForKey(String key) {
+            } else {
+                VolleyLog.d(
+                        "Could not delete cache entry for key=%s, filename=%s",
+                        e.key, DiskBasedCacheUtility.getFilenameForKey(e.key));
```
Do you need this? You're already in DiskBasedCacheUtility, so I think you can just call #getFilenameForKey

sphill99 replied:
fixed both of these

sophiahaim commented on src/main/java/com/android/volley/toolbox/DiskBasedCacheUtility.java:
```diff
@@ -37,6 +42,85 @@ static String getFilenameForKey(String key) {
+            iterator.remove();
+            prunedFiles++;
+
+            if (totalSize < maxCacheSizeInBytes * DiskBasedCacheUtility.HYSTERESIS_FACTOR) {
```
Same here – I think you can just use HYSTERESIS_FACTOR directly

sophiahaim commented on src/main/java/com/android/volley/toolbox/CacheHeader.java:
```diff
@@ -143,4 +145,31 @@ boolean writeHeader(OutputStream os) {
+
+    /** Gets the size of the header in bytes */
+    int getHeaderSize() throws IOException {
+        int x = 44; // 4 longs, 3 ints
```
make this a CONSTANT

Also, use a clearer variable name instead of x – maybe size?

sphill99 commented on src/main/java/com/android/volley/toolbox/DiskBasedCacheUtility.java:
```diff
@@ -104,6 +188,17 @@ static void writeString(OutputStream os, String s) throws IOException {
         os.write(b, 0, b.length);
     }
 
+    static void writeString(ByteBuffer buffer, String s) throws IOException {
```
added @nullable in all the necessary spots.

jbir789 commented on src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java:
```diff
@@ -91,8 +96,77 @@ public void failed(Throwable exc, Void v) {
+
+    /** Re-initialize the cache if the directory was deleted. */
+    private void initializeIfRootDirectoryDeleted() {
+        if (!mRootDirectorySupplier.get().exists()) {
```
nit: what do you think of returning early instead? I find that it improves readability as there is less nested code

```java
if (mRootDirectorySupplier.get().exists()) {
  return;
}

// Else do your work here
```

sphill99 replied:
fixed

jbir789 commented on src/main/java/com/android/volley/toolbox/CacheHeader.java:
```diff
@@ -143,4 +145,31 @@ boolean writeHeader(OutputStream os) {
+                x += etag.getBytes("UTF8").length;
+            }
+        } catch (UnsupportedEncodingException e) {
+            VolleyLog.e(e.toString());
```
Could you log the exception with a message instead?

```java
VolleyLog.e(e, "Your message here")
```

sphill99 replied:
fixed

jbir789 commented on src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java:
```diff
@@ -91,8 +96,77 @@ public void failed(Throwable exc, Void v) {
+        // deleted, then skip writing the entry in the first place, as this is just churn.
+        // Note that we don't include the cache header overhead in this calculation for simplicity,
+        // so putting entries which are just below the threshold may still cause this churn.
+        if (mTotalSize + entry.data.length > mMaxCacheSizeInBytes
+                && entry.data.length
+                        > mMaxCacheSizeInBytes * DiskBasedCacheUtility.HYSTERESIS_FACTOR) {
```
Should these `>` signs be replaced with `>=`? Looking at `pruneIfNeeded`, my understanding is that if the new size is bigger or equal to the max size, the cache would be pruned.

sphill99 replied:
I didn't write these checks, so I'm not totally sure why this is, but it does look like the cache is pruned if the total size if the same as the max size. Would be want to change the check, or change in pruneIfNeeded?

jbir789 replied:
To keep it functionally the same, I'd update these. Note that if you have a method to check those conditions (see the other comment), we wouldn't need to update them because the shared method would take care of keeping the check consistent for us.

jbir789 commented on src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java:
```diff
@@ -91,8 +96,77 @@ public void failed(Throwable exc, Void v) {
+        // deleted, then skip writing the entry in the first place, as this is just churn.
+        // Note that we don't include the cache header overhead in this calculation for simplicity,
+        // so putting entries which are just below the threshold may still cause this churn.
+        if (mTotalSize + entry.data.length > mMaxCacheSizeInBytes
+                && entry.data.length
+                        > mMaxCacheSizeInBytes * DiskBasedCacheUtility.HYSTERESIS_FACTOR) {
```
It's a good idea to check for this. However, we now have duplicate code in `pruneIfNeeded` and here, in the sense we're checking the same conditions in 2 different places. This is error prone: in the future, if one condition is updated and someone forgets to update the other, it would result in the 2 checks being out of sync and we could end up with not adding entries to the cache when it would have been totally fine.

Instead, I suggest to have your conditions defined in a shared place, most likely in `DiskBasedCacheUtility`. It's OK to split them into 2 methods (one for checking the max cache size, and one with the hysteresis factor), but at least you'd have them defined in one place which will prevent them from being out of sync. There would still be a risk that `pruneIfNeeded` is updated and its logic doesn't correspond to the check here, so it's not perfect, but the chances of getting out of sync are decreased.

sphill99 replied:
Yeah, I agree with that. I just made those changes.

jbir789 commented on src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java:
```diff
@@ -91,8 +96,77 @@ public void failed(Throwable exc, Void v) {
+                    /* attachment= */ null,
+                    new CompletionHandler<Integer, Void>() {
+                        @Override
+                        public void completed(Integer integer, Void aVoid) {
```
could you find a better name for `integer` and `aVoid` to be more descriptive? As is, I don't know what they stand for when reading the code.

If they're ignored, you can name them `ignore`, but prefer a descriptive name, especially for the integer.

sphill99 replied:
I'll change these

jbir789 commented on src/main/java/com/android/volley/toolbox/DiskBasedCacheUtility.java:
```diff
@@ -37,6 +42,85 @@ static String getFilenameForKey(String key) {
+     * @return A long to update the total size.
+     */
+    static long putEntry(
+            String key, CacheHeader entry, long totalSize, Map<String, CacheHeader> entries) {
```
I understand why you're passing `entries` here, now that it's static, but keep in mind that it's generally not a good practice to modify the arguments of a method, as they're interpreted as an input to the function and not something that will be modified. The user will have to check the doc (which btw should reflect that `entries` will be updated`) to fully understand what the method does (and while it's always a good idea to do so, you can't guarantee future developers will do so, and it's always better to go with self-explicit methods).

I could be convinced that it's fine to keep this as is because it's internal, but we would need a more descriptive documentation. Alternatively, you could keep the method to compute the new total size, and do the `entries.put(key, entry)` at the caller site instead. I would personally prefer this approach.

sphill99 replied:
Would this be true for pruneIfNeeded as well, since it also changes the entries map? And there will probably be a couple more methods like this (removeEntry, etc.) eventually. Since this would duplicate a lot of code, do you still think that this would be the better route?

Also, we could have the methods return both the entries map and the updated totalSize, then set it in the caller. What're your thoughts?

jbir789 replied:
I thought about returning both, but I am not sure it would make the code more readable, so we can stick to this if you think it's better this way.

jbir789 commented on src/main/java/com/android/volley/toolbox/DiskBasedCacheUtility.java:
```diff
@@ -37,6 +43,95 @@ static String getFilenameForKey(String key) {
+        return new File(rootDirectorySupplier.get(), getFilenameForKey(key));
+    }
+
+    static boolean checkPrune(long totalSize, int dataLength, int maxCacheSize) {
```
nit: the method name isn't super descriptive. What do you think of `wouldExceedCacheSize`?

sphill99 replied:
yeah, I like that more

jbir789 commented on src/main/java/com/android/volley/toolbox/DiskBasedCacheUtility.java:
```diff
@@ -37,6 +43,95 @@ static String getFilenameForKey(String key) {
+            iterator.remove();
+            prunedFiles++;
+
+            if (totalSize < maxCacheSizeInBytes * HYSTERESIS_FACTOR) {
```
Can you use the new function here?

sphill99 replied:
I think we can, but should change the methods to <= then. As otherwise it wouldn't cover the same cases.

jbir789 commented on src/main/java/com/android/volley/toolbox/DiskBasedCacheUtility.java:
```diff
@@ -37,6 +43,95 @@ static String getFilenameForKey(String key) {
+            int maxCacheSizeInBytes,
+            Map<String, CacheHeader> entries,
+            FileSupplier rootDirectorySupplier) {
+        if (totalSize < maxCacheSizeInBytes) {
```
Can you use the new function here?

jbir789 commented on src/main/java/com/android/volley/toolbox/DiskBasedCacheUtility.java:
```diff
@@ -37,6 +43,95 @@ static String getFilenameForKey(String key) {
+        return new File(rootDirectorySupplier.get(), getFilenameForKey(key));
+    }
+
+    static boolean checkPrune(long totalSize, int dataLength, int maxCacheSize) {
```
You could make the function more versatile and really ensure all the checks are in sync by having it take a `newTotalSize` instead of both a `totalSize` and `dataLength`

jbir789 commented on src/main/java/com/android/volley/toolbox/DiskBasedCacheUtility.java:
```diff
@@ -37,6 +43,95 @@ static String getFilenameForKey(String key) {
+        return (totalSize + dataLength > maxCacheSize);
+    }
+
+    static boolean checkWouldDelete(int dataLength, int maxCacheSize) {
```
Could you find a more descriptive name here as well?

sphill99 replied:
changed

sphill99 commented:
I believe I have addressed all comments on this PR

jbir789 commented on src/main/java/com/android/volley/toolbox/DiskBasedCacheUtility.java:
```diff
@@ -37,6 +43,95 @@ static String getFilenameForKey(String key) {
+
+    static boolean wouldExceedCacheSize(long newTotalSize, int maxCacheSize) {
+        return (newTotalSize >= maxCacheSize);
+    }
```
I think it's safer to keep > rather than >= because that's what would have result in a prune, what do you think? (Same below)

sphill99 replied:
yeah, I think that makes sense. I'll change it back.

sphill99 replied:
I changed it in the method, but this lead to a test failing. In pruneIfNeeded, it was originally checking if totalSize <= maxCacheSize. To use the method here, I added 1 to totalSize, so to make the check the same as before.

jbir789 replied:
Should we update the test instead, by any chance?

joebowbeer replied:
Agree. This code originated from

```
        if (mTotalSize + entry.data.length > mMaxCacheSizeInBytes
                && entry.data.length
                        > mMaxCacheSizeInBytes * DiskBasedCacheUtility.HYSTERESIS_FACTOR) {
            return;
        }
```

jbir789 replied:
`DiskBasedCache` had the following check in `#put`:

```java
        if (mTotalSize + entry.data.length > mMaxCacheSizeInBytes
                && entry.data.length
                        > mMaxCacheSizeInBytes * DiskBasedCacheUtility.HYSTERESIS_FACTOR) {
          return;
        }
```

sphill99 replied:
I don't think it super matters anyway. Since the check in put doesn't count the size of the CacheHeader (aside from the data itself). So if it was already really close to the max, it may get put and then pruned. I switched it to >= since that does make more sense (and then holds for the check in pruneIfNeeded), but I don't think it makes a big difference either way.

jbir789 replied:
Sounds good, thanks!

joebowbeer commented on src/main/java/com/android/volley/toolbox/DiskBasedCacheUtility.java:
```diff
@@ -37,6 +43,95 @@ static String getFilenameForKey(String key) {
+            int maxCacheSizeInBytes,
+            Map<String, CacheHeader> entries,
+            FileSupplier rootDirectorySupplier) {
+        if (!wouldExceedCacheSize(totalSize + 1, maxCacheSizeInBytes)) {
```
This `+1` needs a comment here, at least, and maybe in the method as well. How is the caller suppose to know they should add one?

sphill99 replied:
Yeah, I'll add comments explaining this. I did this to essentially make the less than check, a less than or equal to check to match the original code.

joebowbeer replied:
The original code here is:

```
if (mTotalSize < mMaxCacheSizeInBytes) {
            return;
        }
```

The comparison here is would be less than, rather than not would exceed. I think the function call isobscuring the logic, and it's not hiding anything data.

The test is correct.

jbir789 commented on src/main/java/com/android/volley/toolbox/DiskBasedCacheUtility.java:
```diff
@@ -37,6 +43,95 @@ static String getFilenameForKey(String key) {
+    }
+
+    static boolean wouldExceedCacheSize(long newTotalSize, int maxCacheSize) {
+        return (newTotalSize >= maxCacheSize);
```
nit: the parens aren't needed

jbir789 commented on src/main/java/com/android/volley/toolbox/DiskBasedCacheUtility.java:
```diff
@@ -37,6 +43,95 @@ static String getFilenameForKey(String key) {
+            iterator.remove();
+            prunedFiles++;
+
+            if (!isDataTooLarge((int) totalSize, maxCacheSizeInBytes)) {
```
Casting a `long` to `int` is a lossy conversion: if the value of the long needs more than 32 bits, some of them will be discarded to fit in the `int`.

It'd be safer for `isDataTooLarge` to take a `long` instead (so you'd be casting an `int` to `long`, which is fine)

sphill99 replied:
oh yeah, that makes sense. thanks!

jbir789 commented on src/main/java/com/android/volley/toolbox/DiskBasedCacheUtility.java:
```diff
@@ -37,6 +43,95 @@ static String getFilenameForKey(String key) {
+        return new File(rootDirectorySupplier.get(), getFilenameForKey(key));
+    }
+
+    static boolean wouldExceedCacheSize(long newTotalSize, int maxCacheSize) {
```
nit: make `maxCacheSize` a long to be consistent with the total size

joebowbeer replied:
The term `exceed` in this method name is misleading as implemented, but I think exceed is correct.

I'll collect comments on these two methods below.

sophiahaim commented on src/main/java/com/android/volley/AsyncCache.java:
```diff
@@ -24,5 +24,12 @@
     public abstract void get(String key, OnGetCompleteCallback callback);
 
+    public interface OnPutCompleteCallback {
+        /** Invoked when the put to the cache is complete */
```
super nit: add a period

sophiahaim commented on src/main/java/com/android/volley/toolbox/CacheHeader.java:
```diff
@@ -14,6 +16,9 @@
     /** Magic number for current version of cache file format. */
     private static final int CACHE_MAGIC = 0x20150306;
 
+    /** Bits required to write 4 longs and 3 ints */
```
Same here – add a period

jpd236 replied:
fyi, looks like this is still unaddressed

sphill99 commented:
Had to remove the try-with-resources from the get and put methods, as the channel was closing before the reads and writes were actually finished. Same issue with putting the close function in a finally block. Also had to flip the byte buffer after putting the bytes, and needed to start reading at the beginning of the data.

@joebowbeer currently I split it into two methods, so that the pruneIfNeeded method and the put methods don't get out of sync. Since the pruneIfNeeded method does two separate checks, I split it into two different methods. I could make one wouldBePruned method that could make those two checks, and that would clear up exactly what the check at the beginning of put is doing and move the comment to one spot. I'm not sure if that would settle your concerns with the syncing of these methods.

joebowbeer commented:
> @joebowbeer currently I split it into two methods, so that the pruneIfNeeded method and the put methods don't get out of sync. Since the pruneIfNeeded method does two separate checks, I split it into two different methods. I could make one wouldBePruned method that could make those two checks, and that would clear up exactly what the check at the beginning of put is doing and move the comment to one spot. I'm not sure if that would settle your concerns with the syncing of these methods.

1. The quick-return check in `pruneIfNeeded` should be `totalCacheSize <= maxCacheSize`. It should check if the cache does not violate its size restriction. If not then quit now before pruning. (The missing `=` sign was introduced a couple years ago when the size check was moved from pre-write to post-write.)

2. Aside: I don't think the expression above can be made any clearer. The variables are clearly named. Everyone understands the `<=` operator. Using a method just obfuscates it, esp. if you're passing-in all the arguments, and esp. if the method doesn't even do what it proclaims to do. It'sessentially a glorified `greaterThanOrEqual` method that is misnamed `greaterThan`.

3. Yes, I think `wouldBePruned` is a good name for the method that can be shared between its two users. Please unit test this, particularly around the int/long boundary and around the maxCacheSize boundary.

jbir789 commented on src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java:
```diff
@@ -76,12 +83,22 @@ public void completed(Integer result, Void v) {
+                            try {
+                                afc.close();
+                            } catch (IOException e) {
+                                VolleyLog.e(e, "Failed to close channel");
```
super small nit: having more descriptive error logs can be helpful to narrow down where the error comes from. This is why I'd suggest to say something like `Failed to close channel after successful read` here, and `Failed to close channel after failed read attempt` below?

jbir789 commented on src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java:
```diff
@@ -91,8 +108,92 @@ public void failed(Throwable exc, Void v) {
+                            try {
+                                afc.close();
+                            } catch (IOException e) {
+                                VolleyLog.e(e, "Failed to close channel");
```
Same for this log

jbir789 commented on src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java:
```diff
@@ -91,8 +108,92 @@ public void failed(Throwable exc, Void v) {
+                            try {
+                                afc.close();
+                            } catch (IOException e) {
+                                VolleyLog.e(e, "Failed to close channel");
```
And this one

jbir789 commented on src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java:
```diff
@@ -91,8 +108,92 @@ public void failed(Throwable exc, Void v) {
+                        }
+
+                        @Override
+                        public void failed(Throwable throwable, Void aVoid) {
```
nit: can you rename `aVoid` to `ignore`?

jbir789 commented on src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java:
```diff
@@ -76,12 +83,22 @@ public void completed(Integer result, Void v) {
+                            }
                             callback.onGetComplete(null);
                         }
                     });
```
Should we close the afc in the catch below as well? If we failed to read the file, and respond to the callback, I don't expect we still need the channel to be open (assuming we were able to open it)

sphill99 replied:
It doesn't let me add it there, since the channel is defined in the try block. And I can't declare the channel outside of it, since it needs to be final. Is there any way around this? I think we should be closing it, since IOExceptions can be thrown between opening the channel and performing the read/write.

joebowbeer replied:
Strictly speaking the final catch(IOE) should close the channel if it is open. To better isolate the different failures, I would add a try/finally after opening the channel. Where the finally block closes the channel. Then outer try/catch can just report the error. (See pattern in DiskBasedCache.)

OTOH, the afc.read can throw IOE (NonReadableChannelException) only if channel is not open for reading, so you might never need to close in this exceptional case, but comments are needed to explain, and may also be needed to disable static checks.

sphill99 replied:
I tried it with a finally block, but with an AsynchronousFileChannel, it will get closed before finishing the read/write, and then cause an error. And there can be an IOException thrown while writing the CacheHeader (UnsupportedEncodingException), so that would be a case where the channel wouldn't actually be closed. I think closing the channel in the catch block may be the only way to do it.

joebowbeer replied:
@sphill99  I see what you mean now

jbir789 commented on src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java:
```diff
@@ -91,8 +108,92 @@ public void failed(Throwable exc, Void v) {
+                        }
+                    });
+        } catch (IOException e) {
+            boolean deleted = file.delete();
```
Same here, should we close the channel if it was open?

jpd236 replied:
It seems like we should try to close it before trying to delete the file. I would imagine that deleting a file and then calling close() could cause the file to reappear, but not sure.

sphill99 replied:
if the close fails, would we still try to delete the file?

joebowbeer commented:
@jbir789 My primary request is to consolidate the `wouldBePruned` method, for the comment if for no other reason.

As for the test in `pruneIfNeeded`, it is potentially a different test, since it is performed **after** the cache is written, so costs are different. Also, consolidating `wouldBePruned` may reduce the number of call sites below three (and so rule-of-three no longer applies?). More to the point, I think that creating a method for `>` is a false economy. Regardless of the shared method and what its name implies, the right arguments need to be supplied in the right order. Adding a method makes the code more difficult to review at the local level, e.g., if a change is introduced in only one location where the method body is not visible. I don't think it helps, but I can't object if the method implementation matches its name and if the code is effectively tested.

I do think the quick-return test in `pruneIfNeeded` should `<=`. That does not need to be part of this PR, but if the code is being changed anyway...

sphill99 commented:
I believe I've now addressed all comments on this PR. Let me know if I missed any, or didn't make the necessary changes.

jpd236 commented on src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java:
```diff
@@ -76,23 +87,129 @@ public void completed(Integer result, Void v) {
+                        @Override
+                        public void completed(Integer resultLen, Void ignore) {
+                            header.size = resultLen;
+                            mTotalSize =
```
It looks a little peculiar to have:

mTotalSize = x;
mTotalSize = y;

rather than just
x;
mTotalSize = y;

since the first set is immediately clobbered by the second.

That seems worth fixing, but it does actually raise a question that I don't think we've considered yet. What thread safety guarantees should this cache provide? The old cache had "synchronized" keywords over all public API methods, and since the methods are blocking, that ensures thread safety by ensuring only one operation can occur at a time on the entire cache. The new implementation doesn't appear to do anything to address thread safety, and synchronized keywords wouldn't be enough since the disk operations are now asynchronous. I also think we need to be careful about when we try to grab the lock - if we try to grab it before triggering an async operation, and the lock will be blocked on some other async operation, then we effectively make the non-blocking API blocking. So we'll need to think through solutions here.

I wouldn't block this PR on it - just something we'll have to address in a follow-up.

joebowbeer replied:
The first value is used to compute the second.

jpd236 commented on src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java:
```diff
@@ -76,23 +87,129 @@ public void completed(Integer result, Void v) {
+                            try {
+                                afc.close();
+                            } catch (IOException e) {
+                                VolleyLog.e(e, "Failed to close channel after successful write");
```
I don't think the write can be considered successful if the close fails - e.g. I don't think we should update total size or perform pruning until the close completes, and then only if it's successful.

jpd236 commented on src/main/java/com/android/volley/toolbox/DiskBasedCache.java:
```diff
@@ -241,8 +236,10 @@ public synchronized void put(String key, Entry entry) {
             e.size = file.length();
-            putEntry(key, e);
-            pruneIfNeeded();
+            mTotalSize = DiskBasedCacheUtility.putEntry(key, e, mTotalSize, mEntries);
```
remove "mTotalSize = " - no need to set this twice, and the parent method and I believe all other reads of mTotalSize are synchronized so mTotalSize shouldn't be observable between these two sets.

joebowbeer replied:
Disagree. The updated value of mTotalSize is passed to pruneIfNeeded. (I hope a unit test would catch this...)

By the way, having to pass all of these values to the tool method, in the right order, and then update the instance fields, seems brittle. Can't OOP offer more assistance more here?

sphill99 replied:
but it is passed into the next method, so doesn't it need to be updated there?

sphill99 replied:
Yeah, I was thinking about better ways to do this. I think it's good to share the code, since it'd be a lot of code to be duplicated. I don't think this is terrible, since the methods take in all different types, which means it'll tell you if the arguments aren't in the right order. The only other method I could think of is passing in the Cache itself. And then making those variables accessible. But I think that may cause more issues.

jpd236 replied:
Yeah, sorry, didn't realize it was passed from the first method to the second. Never mind here.

jpd236 commented on src/main/java/com/android/volley/toolbox/DiskBasedCacheUtility.java:
```diff
@@ -37,6 +43,120 @@ static String getFilenameForKey(String key) {
+    }
+
+    static boolean wouldExceedCacheSize(long newTotalSize, long maxCacheSize) {
+        return (newTotalSize > maxCacheSize);
```
nit: these parentheses are unnecessary

jpd236 commented on src/main/java/com/android/volley/toolbox/DiskBasedCacheUtility.java:
```diff
@@ -136,4 +281,18 @@ static void writeHeaderList(List<Header> headers, OutputStream os) throws IOExce
+        int bytes = 4;
+
+        for (Header header : headers) {
+            bytes += header.getName().getBytes("UTF-8").length;
```
we use UTF-8 here and UTF8 elsewhere. Can we be consistent? Why do we check for UnsupportedEncodingException there and not here?

sphill99 replied:
Would it make sense to require newer api for these methods, then use StandardCharsets.UTF-8? It's a nio class, and never will have an UEE.

joebowbeer replied:
UTF-8 is the default and is always available

sphill99 replied:
I was getting compilation errors unless I specified UTF-8

jpd236 replied:
I don't think we can use StandardCharsets here because this code is shared with the DiskBasedCache code path, which needs more backwards compatibility. Just want to make sure we're consistent with the naming.

sphill99 replied:
Okay, I just changed this back. They should all now be "UTF-8"

sphill99 commented on src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java:
```diff
@@ -76,23 +87,129 @@ public void completed(Integer result, Void v) {
+                    new CompletionHandler<Integer, Void>() {
+                        @Override
+                        public void completed(Integer resultLen, Void ignore) {
+                            header.size = resultLen;
```
yeah, I think that'd be a good check to make

sphill99 commented on src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java:
```diff
@@ -76,23 +87,129 @@ public void completed(Integer result, Void v) {
                                 return;
                             }
                             byte[] data = buffer.array();
+                            try {
```
I added a helper method to make this cleaner. I wasn't sure if this was exactly what you were intending though. Let me know if I should change anything about this.

jpd236 commented on src/main/java/com/android/volley/AsyncCache.java:
```diff
@@ -24,5 +24,12 @@
+        void onPutComplete();
+    }
+
+    public abstract void put(String key, Cache.Entry entry, OnPutCompleteCallback callback);
```
could you add Javadoc for this (like get)

sphill99 replied:
added javadoc here

jpd236 commented on src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java:
```diff
@@ -80,19 +92,119 @@ public void completed(Integer result, Void v) {
+                            if (closeChannel(afc, "completed read")) {
+                                if (resultLen != size) {
+                                    VolleyLog.e(
+                                            "File changed while reading: %s",
```
this is a write, correct?

Also, I think we should delete the file in this case, since it wouldn't have valid contents.

jpd236 commented on src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java:
```diff
@@ -80,19 +92,119 @@ public void completed(Integer result, Void v) {
+                    new CompletionHandler<Integer, Void>() {
+                        @Override
+                        public void completed(Integer resultLen, Void ignore) {
+                            if (closeChannel(afc, "completed read")) {
```
write

jpd236 commented on src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java:
```diff
@@ -80,19 +92,119 @@ public void completed(Integer result, Void v) {
+                    new CompletionHandler<Integer, Void>() {
+                        @Override
+                        public void completed(Integer resultLen, Void ignore) {
+                            if (closeChannel(afc, "completed read")) {
```
and I think we should delete after a failed close as well, yes. in the existing implementation we delete in the event of any IO failure.

sphill99 replied:
fixed

jpd236 commented on src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java:
```diff
@@ -80,19 +92,119 @@ public void completed(Integer result, Void v) {
+        initialize();
+    }
+
+    private boolean closeChannel(AsynchronousFileChannel afc, String endOfMessage) {
```
`@Nullable` afc

jpd236 commented on src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java:
```diff
@@ -80,19 +92,119 @@ public void completed(Integer result, Void v) {
+        initialize();
+    }
+
+    private boolean closeChannel(AsynchronousFileChannel afc, String endOfMessage) {
```
consider a quick Javadoc, since the meaning of the return value is perhaps non-obvious.

sphill99 replied:
added javadoc

sphill99 commented:
I believe I've addressed all comments currently on this PR.

jbir789 commented on src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java:
```diff
@@ -80,19 +93,136 @@ public void completed(Integer result, Void v) {
+            afc.close();
+            return true;
+        } catch (IOException e) {
+            VolleyLog.e(e, "failed to close file after " + endOfMessage);
```
nit: use the String format type of log: `log(e. "failed to close file after %s", endOfMessage)`

jpd236 commented on src/main/java/com/android/volley/toolbox/DiskBasedCache.java:
```diff
@@ -83,7 +81,8 @@ public File get() {
      */
-    public DiskBasedCache(FileSupplier rootDirectorySupplier, int maxCacheSizeInBytes) {
+    public DiskBasedCache(
+            DiskBasedCacheUtility.FileSupplier rootDirectorySupplier, int maxCacheSizeInBytes) {
```
Actually, this is an unsafe API change. This is a public constructor (so Volley clients may have used this), so we can't change the signature. (Moreover, DiskBasedCacheUtility is package private, so external callers can't even access it).

Unfortunately I think we need to leave FileSupplier here and duplicate it for the Async layer.

sphill99 replied:
This should be fixed now that FileSupplier is duplicated in DiskBasedCache

jpd236 commented on src/main/java/com/android/volley/toolbox/DiskBasedCache.java:
```diff
@@ -255,31 +252,28 @@ public synchronized void put(String key, Entry entry) {
 
+    /** Represents a supplier for {@link File}s. */
+    public interface FileSupplier extends DiskBasedCacheUtility.FileSupplier {
+        @Override
```
Is this necessary? I think you can just do "extends DiskBasedCacheUtility.FileSupplier { }" since you're not changing anything about the underlying method.

sphill99 replied:
Oh, I'll move the File Supplier. And you're right,  that isn't necessary,  I'll change that as well

sphill99 replied:
I had to change the name to FileSupplierParent, as otherwise, it couldn't be extended by FileSupplier

jpd236 replied:
Can you call it FileSupplier and do "public interface FileSupplier implements com.google.android.volley.toolbox.FileSupplier"?

sphill99 replied:
Yep, just tried that and it works!
